marklogic-data-hub/src/server-side/dhf.sjs

```diff
@@ -0,0 +1,218 @@
+/**
+  Copyright 2012-2018 MarkLogic Corporation
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+*/
+
+const flowlib = require("/MarkLogic/data-hub-framework/impl/flow-lib.sjs");
+const tracelib = require("/MarkLogic/data-hub-framework/impl/trace-lib.sjs");
+
+const _requireCache = {};
+
+/**
+ : Runs a given function as a plugin. This method provides
+ : tracing around your function. Tracing will catch uncaught
+ : exceptions and log them into the traces database.
+ :
+ : @param context - the context for this plugin
+ : @param func - the function to run
+ : @return - returns whatever your function returns
+ */
+function run(context, func)
+{
+  let label = context.label;
+
+  if (label) {
+    tracelib.setPluginLabel(label);
+  } else {
+    fn.error(xs.QName("CONTEXT_MISSING_LABEL"), "Your context object is missing a label");
+  }
+
+  tracelib.resetPluginInput();
+
+  let inputs = context.inputs
+  for (let key in inputs) {
+    tracelib.setPluginInput(key, inputs[key]);
```

Comment:
``if (obj.hasOwnProperty(prop)) {`` needed here?

Reply:
nope. that blows up.

Reply:
I meant inputs.hasOwnProperty. Only relevant if inputs might inherit properties from somewhere.

Reply:
yeah. I fixed it to the right thing and it still blew up.